### DISCUSSION TITLE
Allow for non-boolean singletons in if statements

### DISCUSSION THREAD

**Discussion by llfletch** [dotnet/csharplang#1463] [Language Ideas]
What if we were to allow if statements to support non-Boolean arguments if the arguments were singletons (the only argument in the if statement)?  The main problem with using non-Boolean arguments in if statements seems to be having an assignment statement in the if statement such that the user doesn't realize that it is an assignment statement rather than a comparison.  With singleton arguments, though, this isn't a problem and it could significantly shorten many if statements.  For example:
`if (variable != null)`  becomes `if (variable)`

When evaluated in an if statement, strings could be true if they have data and false if they are empty or null.  Also, structures and object could be true if they contain any public data or false if they contain only the starting values (zero, false, null, etc.).

**Comment by svick** [comment]
> When evaluated in an if statement, strings could be true if they have data and false if they are empty or null. Also, structures and object could be true if they contain any public data or false if they contain only the starting values (zero, false, null, etc.).

I think this would only lead to confusion about what exactly does `if (x)` do.

In this specific case, I believe clarity is more important than brevity.

---

Also, it's not clear to me what a "singleton" is, since an `if` statement always takes only a single expression.

**Comment by bondsbw** [comment]
Similar:  #871, #1410

**Comment by jnm2** [comment]
Also see the discussion in https://github.com/dotnet/csharplang/issues/196, https://github.com/dotnet/csharplang/issues/438, https://github.com/dotnet/csharplang/issues/1455.

**Comment by FurkanKambay** [comment]
If you want this for some classes you created, you can add an implicit operator for `bool` to your class:

```C#
public static implicit operator bool(Foo foo)
{
    return foo.IsValid; // this shows that it does not have to be a null check.
}
```

edit: or `true` and `false` operators if you don't want implicit casting. which are implemented like this: (you have to implement both)
```C#
public static bool operator true(Foo foo) => foo.IsValid;
public static bool operator false(Foo foo) => foo.IsValid;
```

**Comment by Thaina** [comment]
@FurkanKambay I think we could `public static explicit operator bool(Foo foo)`

**Comment by FurkanKambay** [comment]
@Thaina well that would require us to write `if ((bool)foo) { }` instead of `if (foo) { }` but I guess that has an advantage of being explicit.

**Comment by Thaina** [comment]
@FurkanKambay It seem I misunderstand your sentence `if you don't want implicit casting`
